pkg/streamingpromql/pooling/limiting_pool_test.go

```diff
@@ -0,0 +1,184 @@
+	put(s100)
+	require.Equal(t, 2*elementSize, pool.CurrentEstimatedMemoryConsumptionBytes)
+	require.Equal(t, 130*elementSize, pool.PeakEstimatedMemoryConsumptionBytes)
+
+	// Get another slice from the pool that doesn't take us over the previous peak.
+	s5, err := get(5)
+	require.NoError(t, err)
+	require.Equal(t, 8, cap(s5))
+	require.Equal(t, 10*elementSize, pool.CurrentEstimatedMemoryConsumptionBytes)
+	require.Equal(t, 130*elementSize, pool.PeakEstimatedMemoryConsumptionBytes)
+
+	// Get another slice from the pool that does take us over the previous peak.
+	s200, err := get(200)
+	require.NoError(t, err)
+	require.Equal(t, 256, cap(s200))
+	require.Equal(t, 266*elementSize, pool.CurrentEstimatedMemoryConsumptionBytes)
+	require.Equal(t, 266*elementSize, pool.PeakEstimatedMemoryConsumptionBytes)
+
+	// Ensure we handle nil slices safely.
+	put(nil)
+	require.Equal(t, 266*elementSize, pool.CurrentEstimatedMemoryConsumptionBytes)
+	require.Equal(t, 266*elementSize, pool.PeakEstimatedMemoryConsumptionBytes)
+}
+
+func TestLimitingPool_Limited(t *testing.T) {
+	t.Run("[]promql.FPoint", func(t *testing.T) {
+		pool := NewLimitingPool(11 * FPointSize)
+		testLimitedPool(t, pool.GetFPointSlice, pool.PutFPointSlice, pool, FPointSize)
+	})
+
+	t.Run("[]promql.HPoint", func(t *testing.T) {
+		pool := NewLimitingPool(11 * HPointSize)
+		testLimitedPool(t, pool.GetHPointSlice, pool.PutHPointSlice, pool, HPointSize)
+	})
+
+	t.Run("promql.Vector", func(t *testing.T) {
+		pool := NewLimitingPool(11 * VectorSampleSize)
+		testLimitedPool(t, pool.GetVector, pool.PutVector, pool, VectorSampleSize)
+	})
+
+	t.Run("[]float64", func(t *testing.T) {
+		pool := NewLimitingPool(11 * Float64Size)
+		testLimitedPool(t, pool.GetFloatSlice, pool.PutFloatSlice, pool, Float64Size)
+	})
+
+	t.Run("[]bool", func(t *testing.T) {
+		pool := NewLimitingPool(11 * BoolSize)
+		testLimitedPool(t, pool.GetBoolSlice, pool.PutBoolSlice, pool, BoolSize)
+	})
+}
+
+func testLimitedPool[E any, S ~[]E](t *testing.T, get func(int) (S, error), put func(S), pool *LimitingPool, elementSize uint64) {
+	// This method assumes that pool has been created with a limit set to 11x elementSize
+
+	// Get a slice from the pool beneath the limit.
+	s7, err := get(7)
+	require.NoError(t, err)
+	require.Equal(t, 8, cap(s7))
+	require.Equal(t, 8*elementSize, pool.CurrentEstimatedMemoryConsumptionBytes)
+	require.Equal(t, 8*elementSize, pool.PeakEstimatedMemoryConsumptionBytes)
+
+	// Get another slice from the pool beneath the limit.
+	s1, err := get(1)
+	require.NoError(t, err)
+	require.Equal(t, 1, cap(s1))
+	require.Equal(t, 9*elementSize, pool.CurrentEstimatedMemoryConsumptionBytes)
+	require.Equal(t, 9*elementSize, pool.PeakEstimatedMemoryConsumptionBytes)
+
+	// Return a slice to the pool.
+	put(s1)
+	require.Equal(t, 8*elementSize, pool.CurrentEstimatedMemoryConsumptionBytes)
+	require.Equal(t, 9*elementSize, pool.PeakEstimatedMemoryConsumptionBytes)
+
```

Comment:
(nit) it would be good to check if we can get another slice below the limit after putting a slice back into a pool. Specifically where if we did not put a slice back in the pool we would be over the limit. ie ensure that we are properly discounting the used memory when putting back to the pool.
(Granted it's done on line 143, so this is about making it more explicit in case the tests or slice sizes change in the future etc)

Reply:
Makes sense to me, I've expanded the tests in 86649ca.